pkg/channel/fanout/fanout_message_handler.go

```diff
@@ -120,46 +122,92 @@ func SubscriberSpecToFanoutConfig(sub eventingduckv1.SubscriberSpec) (*Subscript
 
 func createMessageReceiverFunction(f *MessageHandler) func(context.Context, channel.ChannelReference, binding.Message, []binding.Transformer, nethttp.Header) error {
 	if f.config.AsyncHandler {
-		return func(ctx context.Context, _ channel.ChannelReference, message binding.Message, transformers []binding.Transformer, additionalHeaders nethttp.Header) error {
+		return func(ctx context.Context, ref channel.ChannelReference, message binding.Message, transformers []binding.Transformer, additionalHeaders nethttp.Header) error {
 			if len(f.config.Subscriptions) == 0 {
 				// Nothing to do here, finish the message and return
 				_ = message.Finish(nil)
 				return nil
 			}
 
 			parentSpan := trace.FromContext(ctx)
+			te := kncloudevents.TypeExtractorTransformer("")
+			transformers = append(transformers, &te)
 			// Message buffering here is done before starting the dispatch goroutine
 			// Because the message could be closed before the buffering happens
 			bufferedMessage, err := buffering.CopyMessage(ctx, message, transformers...)
 			if err != nil {
 				return err
 			}
+
+			reportArgs := channel.ReportArgs{}
+			reportArgs.EventType = string(te)
+			reportArgs.Ns = ref.Namespace
+
 			// We don't need the original message anymore
 			_ = message.Finish(nil)
-			go func(m binding.Message, h nethttp.Header, s *trace.Span) {
+			go func(m binding.Message, h nethttp.Header, s *trace.Span, r *channel.StatsReporter, args *channel.ReportArgs) {
 				// Run async dispatch with background context.
 				ctx = trace.NewContext(context.Background(), s)
 				// Any returned error is already logged in f.dispatch().
-				_ = f.dispatch(ctx, m, h)
-			}(bufferedMessage, additionalHeaders, parentSpan)
+				dispatchResultForFanout := f.dispatch(ctx, m, h)
+				if dispatchResultForFanout.info != nil && dispatchResultForFanout.info.Time > channel.NoDuration {
+					if dispatchResultForFanout.info.ResponseCode > channel.NoResponse {
+						_ = (*r).ReportEventDispatchTime(&reportArgs, dispatchResultForFanout.info.ResponseCode, dispatchResultForFanout.info.Time)
+					} else {
+						_ = (*r).ReportEventDispatchTime(&reportArgs, nethttp.StatusInternalServerError, dispatchResultForFanout.info.Time)
+
+					}
+				}
+				err = dispatchResultForFanout.err
+				if err != nil {
+					channel.ReportEventCountMetricsForDispatchError(err, f.reporter, reportArgs)
+				} else {
+					if dispatchResultForFanout.info != nil {
+						_ = (*r).ReportEventCount(args, nethttp.StatusAccepted)
+					}
+				}
+
```

Comment:
I meant this code, it's duplicated of lines 195 to 209, can you create a private function for that?

Reply:
Ok now I see what you mean, will do.